### PR TITLE
[OSX] Trackpad gestures - Magnify (Pitch to zoom), Rotate and Swipe

### LINKED ISSUE #6545: [OSX] Support of `Pitch to Zoom` trackpad gesture as a routed event.
**Is your feature request related to a problem? Please describe.**
Pitch to Zoom gesture is a must have feature to support on a macbook. It's used by macbook users everywhere. It's comfortable. Lack of that feature is a big problem in our software.

**Describe the solution you'd like**
I think we need to add an event to handle this in Avalonia, also add a native lib implementation to launch it. 

Apple Documentation - https://developer.apple.com/documentation/appkit/nsresponder/1525862-magnifywithevent

**What do I need?**
I will do a PR by myself. All I need is an API suggestion/approval. Where this event should be, How it should be called, should it be a routed event or not? I think it should be almost like `MouseWheel` routed event and called exactly `PitchToZoom`.

### PR DESCRIPTION
**Pull request by mklts**
## What does the pull request do?
Adds support for three gestures:
Magnify - double Delta
Rotate - double Delta
Swipe - Vector Delta - based on vector you can get a direction

I've added this to PointerDevice because trackpad is a pointer device. I've thought about creation of TrackpadDevice - but this is a bad idea, because gestures above are pointer position dependant.

## How was the solution implemented (if it's not obvious)?
I've subscribed to three native methods in native osx lib, added 3 event types, and integrated them into Interactive. 

## Fixed issues
Fixed #6545 

## WIP
Feature is done but I need to test it on a real app. When this will be done I will fix found issues and remove a WIP flag.

## Win32 and Linux support?
AFAIK it's almost impossible to implement. Also this features works great only on macOS AFAIK. So only macOS users are need it(it's essential for any macOS trackpad user).

### CODE REVIEW

**Comment by Mikolaytis** [comment]
Integrated to our app. Works fine. @danwalmsley please review this :)

**Comment by maxkatz6** [comment]
1. It shouldn't be in InputElement. We have Gestures class with gestures-related events. It's a good place for these events.
2. On other platforms it should be done using simple pointer events, otherwise this feature is 66.6% useless.

**Comment by Mikolaytis** [comment]
@maxkatz6 This is not a touch screen gestures, this is trackpad gestures. You cannot track them using simple pointer events. Also this is an impossible challenge to me to implement Windows Precision Driver support, Apple Bootcamp driver support and Synaptics driver support on Windows and also any linux drivers support. This is a huge task compared to macOS implementation. Windows and Linux trackpads are working too bad. Gestures are not precise = not usable at all, so no one need them (me too). So I can provide only a macOS solution. 

I think this is still a good thing to have a support of this on macOS only for now. Only on macOS this feature is essential. Maybe in the future someone will add support for other platforms if this is will be needed to someone (I think not in any time soon). 

I did not moved this events to Gestures class because I thought it's only for touch screen gestures. 
In my app it doesn't matter where it is. I can try move this events to Gestures class if anyone will agree with this.

**Comment by Mikolaytis** [comment]
So, if I will move this events to Gestures class - they will have these names:
* TrackpadMagnifyGestureEvent
* TrackpadRotateGestureEvent
* TrackpadSwipeGestureEvent
Trackpad word is needed to specify that this events are not a touch screen events.
If everyone agree - I will start the implementation.

**Comment by JanTamis** [comment]
will this work with windows precision trackpads?

**Comment by Mikolaytis** [comment]
@danipen  Nope. I'm waiting for response from Avalonia team on how to make the API.

File: src/Avalonia.Input/InputElement.cs [hunk 1]
```diff
@@ -349,13 +367,43 @@ public InputElement()
...
+        
+        /// <summary>
+        /// Occurs when the user uses swipe gesture on a trackpad and pointer is over the control.
+        /// Works only on macOS.
+        /// </summary>
+        public event EventHandler<PointerTouchPadGestureSwipeEventArgs> PointerTouchPadGestureSwipe
+        {
+            add { AddHandler(PointerTouchPadGestureSwipeEvent, value); }
+            remove { RemoveHandler(PointerTouchPadGestureSwipeEvent, value); }
+        }
```
Comment (Mikolaytis) [inline src/Avalonia.Input/InputElement.cs:406]: Should I remove this?

File: src/Avalonia.Input/InputElement.cs [hunk 2]
```diff
@@ -196,6 +196,24 @@ public class InputElement : Interactive, IInputElement
...
+        /// Defines the <see cref="PointerTouchPadGestureRotate"/> event.
+        /// </summary>
+        public static readonly RoutedEvent<PointerTouchPadGestureRotateEventArgs> PointerTouchPadGestureRotateEvent 
+            = Gestures.PointerTouchPadGestureRotateEvent;
+        
+        /// <summary>
+        /// Defines the <see cref="PointerTouchPadGestureSwipe"/> event.
+        /// </summary>
+        public static readonly RoutedEvent<PointerTouchPadGestureSwipeEventArgs> PointerTouchPadGestureSwipeEvent 
+            = Gestures.PointerTouchPadGestureSwipeEvent;
```
Comment (Mikolaytis) [inline src/Avalonia.Input/InputElement.cs:216]: Should I remove this?

**Comment by Mikolaytis** [comment]
I've renamed event's like @kekekeks requested. Also moved events to Gestures class. I still have two questions above about the API.
Also - @danwalmsley already reviewed native side and said it looks good. 
So if managed API is good - let's merge this PR :)

File: src/Avalonia.Input/Raw/RawPointerEventArgs.cs [hunk 3]
```diff
@@ -21,7 +21,10 @@ public enum RawPointerEventType
         TouchBegin,
         TouchUpdate,
         TouchEnd,
-        TouchCancel
+        TouchCancel,
+        Magnify,
```
Comment (pr8x) [inline src/Avalonia.Input/Raw/RawPointerEventArgs.cs:25]: Would make sense to prefix these with "Gesture" ?

**Comment by pr8x** [comment]
In Windows (10?), gestures are handled via the `WM_GESTURE` event (See https://docs.microsoft.com/en-us/windows/win32/wintouch/windows-touch-gestures-overview#interpreting-windows-touch-gestures). Seems like we can just handle this inside `Avalonia.Win32.WindowImpl.AppWndProc`

**Comment by kekekeks** [comment]
@pr8x how does WM_GESTURE interact with WM_POINTER and WM_TOUCH?

**Comment by pr8x** [comment]
@kekekeks Apparently `WM_TOUCH` and `WM_GESTURE` are obsolete (and [kinda broken](http://the-witness.net/news/2012/10/wm_touch-is-totally-bananas/)) and `WM_POINTER` should be used instead. That is only supported on Windows 8 +, though. Apparently it's possible to do gesture detection with this "Interaction Context" API (see [here](https://www.codeproject.com/Articles/607234/Using-Windows-8-Interaction-Con)), but it definitely seems harder to use than `WM_GESTURE`. 

Good summary: https://lenos19.rssing.com/chan-10584515/all_p6.html

Sooo... actually not that easy :D

File: src/Avalonia.Input/InputElement.cs (hunk 2 above)
Comment (maxkatz6) [reply src/Avalonia.Input/InputElement.cs:216]: Yes, please remove it from InputElement.
It still should be possible to use in from xaml like
```
<Canvas Gestures.PointerTouchPadGestureSwipe="" />
```

**Review by maxkatz6 — APPROVED** [review]

**Pull request merged**